## 第二章 处理文本数据

- 将文本分割为独立的单词词元和子词词元，然后编码为LLM所使用的向量表示

### 2.1 理解词嵌入

- 深度神经网络无法直接处理原始文本，因为文本数据是离散的，无法直接执行数学运算。
- 将数据转化为向量格式的过程称为嵌入embedding，不同的数据类型如文本，图像，音视频等需要使用不同的嵌入模型。
- embedding的本质是将离散对象映射到连续向量空间中的点，从而转化为神经网络可以处理的格式。
- RAG，retrieval-augmented generation，是将句子，段落乃至整个文档嵌入的技术。
- 有多种算法和框架来生成词嵌入，早期流行word2vec。
- 词嵌入的维度dimension可以从一维到数千维不等，更高的维度有助于捕获到更细微的关系，同时牺牲计算效率。
- 大语言模型通常会自行生成嵌入，这些嵌入是输入层的一部分，并且在训练中会进行更新。
- 最小的GPT-2模型参数为1.17亿，嵌入维度为768，GPT-3参数为1750亿，嵌入维度为12288。

### 2.2 文本分词

- 将输入文本分割为独立的词元，是生成嵌入向量所必须的预处理步骤。
- 词元即可以是单个的单词，也可以是诸如标点符号之类的特殊字符。

- 采用Edith Wharton的短篇小说The Verdict为分词文本作为例子。

In [1]:
import urllib.request
url = ("https://raw.githubusercontent.com/rasbt/"
       "LLMs-from-scratch/main/ch02/01_main-chapter-code/"
       "the-verdict.txt")
file_path = "the-verdict.txt"
urllib.request.urlretrieve(url, file_path)

('the-verdict.txt', <http.client.HTTPMessage at 0x75fc6cfe2350>)

In [2]:
with open("the-verdict.txt", "r", encoding="utf-8") as f:
    raw_text = f.read()
print("Total number of characters in the text:", len(raw_text))
print(raw_text[:99])

Total number of characters in the text: 20479
I HAD always thought Jack Gisburn rather a cheap genius--though a good fellow enough--so it was no 


- 将这篇20479个字符的短片小说分割为独立的单词和特殊字符，以便于在后续步骤中将其转化为嵌入向量，进而用于LLM训练。
- 为了获取词元列表，就需要正确的分割文本。
- 何为正确？如移除空格可以减少运算负担，但是会丢失掉空格在上下文中的作用。

- 下面使用正则来示例如何分割

In [3]:
import re
text = "Hello, world. This,  is a test."
result = re.split(r'(\s)', text)
print(result)

['Hello,', ' ', 'world.', ' ', 'This,', ' ', '', ' ', 'is', ' ', 'a', ' ', 'test.']
